### PR TITLE
add error handling in run-caliper.sh and run-caliper.py then retry

### PR DESCRIPTION
**Pull request by ulalaa14**
## Description

Add error handling in command script run-caliper.sh and run-caliper.py. If fail retry for several attempts.
Attempts value stored in config.json
Even fail, the next script still running well.
## Related Issue

https://github.com/Cloud-Prototyping-WS-2019-20/cp_ws_1920/issues/60

## Motivation and Context

Build robustness in the current system by handling the error.

## How Has This Been Tested?

running python3 run-caliper.py
since in the main.py there is still error with the local environment

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist contributor:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by fayanpardo** [comment]
So looking to your code I do not see anything wrong, but caliper is still returning 0 when failing to connect to the blockchain, thus caliper is not being executed again at all. Maybe we should go further, have you seen any possible way even if caliper command is actually benchmarking or not? We can maybe save the console output into a file and search for "benchmark successed" and then we know if it failed or not.

**Comment by ulalaa14** [comment]
hmmm. or probably we delete exit 0 in run-caliper.sh

**Comment by ulalaa14** [comment]
updated

**Review by fayanpardo — CHANGES_REQUESTED** [review]
Now when it fails it seems to retry as expected. However I tried the other way around, tested when caliper works and something seem to be wrong.
First of all it tries again even though it worked.
image

And after trying the number of times specified it says it fails.
image

I just saw your code has no exit(-1) in case of failure.

What I did to test was to execute main.py with your updates.

**Comment by ulalaa14** [comment]
updated it, however works in my local, i have trouble while hitting the node on cloud. need help to test it by others. hopefully it could work well. also i'm using exit 1 because the references i found stated exit 1 means exit because there is an error. nothing specify exit -1.

**Comment by fayanpardo** [comment]
GJ! I tested it and now it does not repeat execution after success. However we may need to improve this script. 
image
Here correctly works as expected.

image
Here caliper clearly fails with 1 gas limit but the command has run successfully so it is detected as correct.

We need to implement another way of detecting errors in caliper. What about this idea: Execute caliper command saving the printed output in a file ("caliper command > execution.log" for example) and then search in this file if contains "Benchmark run successful" which is printed when caliper benchmark finished successfully. If this line is not found then it is a failure and if it is found it is a success.

Any more ideas are welcome!

**Comment by ulalaa14** [comment]
updated it 👍

File: bin/run-caliper.py [hunk 1]
```diff
@@ -41,9 +58,24 @@ def main():
...
+    for i in range(attempt):
+        #running run-caliper.sh with reportname
+        run_command = os.system("bash " + bashfile + " " + reportname)
+        if run_command == 0:
+            print("Running caliper success.")
+            break
+
+        # if it reach the maximum attempt, printing error message
+        if i == attempt-1:
+            print("Meet maximum retry. Running caliper error.")
```
Comment (fayanpardo) [inline bin/run-caliper.py:75]: If this is the final caliper failure error I think it should directly after this message (and inside the if of course) exit(-1). If not, main.py assumes everything ran correctly.

I just tested with all failures and even though caliper.sh returned the error, caliper.py returned success execution even after 3 retries.

**Comment by fayanpardo** [comment]
It does not seem to get the message :/
image

**Comment by ulalaa14** [comment]
updated by fixing typo and add wait command to make sure the first command finished

**Review by fayanpardo — APPROVED** [review]
GJ ! now works

**Pull request merged**